### PR TITLE
Add SLF4J and Logback as logging facilities, closes #38

### LINKED ISSUE #38: Add a logger
- Parent issues: -
- Relevant issues: -
- Estimated effort: 3

* * *

## User story
As a developer I want to have logging facilities so that it's easier to track down bugs

## What
This issue demands a way to log the state of the program

## Why
This issue is needed for the project because it makes it easier to track down bugs.

## How
This feature can be added by implementing a logger, for example SLF4J + a suitable logging framework

## Definition of done
This task can be considered done when

- Specific message can be logged at certain times
- The log messages are always flushed to a log file
- The log file contains all log messages

### PR DESCRIPTION
**Pull request by jverbraeken**
- Relevant Issues: #38 
- Related Pull Requests: -

* * *

## What
This PR adds logging facilities to the project by adding SLF4J, the most popular logging facade for Java, and Logback, the successor of the very popular Log4J and arguably the best free industry-grade logger for Java.

## Why
This PR is needed because we want to be able to trace down where bugs are when the program suddenly crashes and because we want to be able to be notified of warnings that don't crash the program or are useful to the user.

## How
This feature can be tested by just running the app and inspecting the logcat (for example by opening Android Monitor)

## Alternative implementation
SLF4J and Logback are both industry-standard and the most popular logging tools at the moment. People are happy about it and I didn't see any complains about in on the internet. Because logging is a relatively unimportant aspect of our program I don't think it's worth the time to look for even better alternatives as these are a safe bet already.

## Notes
-

### CODE REVIEW

**Comment by jverbraeken** [comment]
@TheBonheurs @ericcornelissen Thanks for the suggestion! I made a document containing some guidelines here: https://docs.google.com/document/d/1g62VM3IgYDweTrvEwqNFFspnIivzrymEIT7UpBHnMeA
However, I think that part of the logging is at places where your intuition thinks it should be. For example, I would log it when a button is pressed, but I wouldn't log it when the user scrolls thought a list of items.

File: app/src/main/res/values/dimens.xml [hunk 1]
```diff
@@ -1,11 +1,9 @@
 <resources>
     <dimen name="fab_margin">16dp</dimen>
-    <dimen name="app_bar_height">180dp</dimen>
-    <dimen name="text_margin">16dp</dimen>
     <dimen name="contact_picture_width">100dp</dimen>
     <dimen name="contact_picture_height">100dp</dimen>
-    <dimen name="contact_picture_name_padding">33dp</dimen>
-    <dimen name="contact_name_margin">8dp</dimen>
+    <!--<dimen name="contact_picture_name_padding">33dp</dimen>-->
+    <!--<dimen name="contact_name_margin">8dp</dimen>-->
```
Comment (ericcornelissen) [inline app/src/main/res/values/dimens.xml]: Why not remove these?

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -2,14 +2,14 @@
...
 
     <string name="title_activity_contact">Contact</string>
-    <string name="msg_email_error">Please enter your email</string>
-    <string name="msg_password_error">Please enter your password</string>
+    <!--<string name="msg_email_error">Please enter your email</string>-->
+    <!--<string name="msg_password_error">Please enter your password</string>-->
     <string name="msg_mismatch">Your password is wrong</string>
     <string name="contact_title">Contact</string>
-    <string name="action_settings">Settings</string>
+    <!--<string name="action_settings">Settings</string>-->
```
Comment (ericcornelissen) [inline app/src/main/res/values/strings.xml]: Why not remove these?

Comment (jverbraeken) [inline app/src/main/res/values/strings.xml]: @ericcornelissen I'm aware that such changes are not preferred in a logging PR, but I had to fix them because otherwise Gradle check would fail each time because of Android Lint. I don't know how the failing code could appear on the master, maybe Travis skips Android Lint, but these issues had to be fixed to let my local check succeed.

**Comment by jverbraeken** [comment]
@ericcornelissen I'm aware that such changes are not preferred in a logging PR, but I had to fix them because otherwise Gradle check would fail each time because of Android Lint. I don't know how the failing code could appear on the master, maybe Travis skips Android Lint, but these issues had to be fixed to let my local check succeed.

**Comment by ericcornelissen** [comment]
@jverbraeken I'm pretty sure it is because you removed [this](https://github.com/ericcornelissen/NervousFish/pull/85/files#diff-39e7d8c00954e920b98e7636f0ac30b2L43)

**Comment by jverbraeken** [comment]
@ericcornelissen Actually not, that line of code was duplicate, because it was also (correctly) in quality.gradle

**Comment by ericcornelissen** [comment]
@jverbraeken are you aware that [you added it](https://github.com/ericcornelissen/NervousFish/pull/85/files#diff-fa247df9ed4713b9e945e19a0d129e6aR88)?

**Comment by jverbraeken** [comment]
@ericcornelissen Ah, I see, thought you was talking about a different line of code. I added it because both issues were not relevant, but although I suppressed both warnings I still got quite some Android Lint warnings that I had to solve before my local build succeeded

**Comment by ericcornelissen** [comment]
@jverbraeken no problem 🙂 still, it would have been better to put it in a different pull request...

**Comment by jverbraeken** [comment]
@ericcornelissen Next time I'll create a different PR to merge into this branch. Is the rest of the PR OK for now?

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java [hunk 3]
```diff
@@ -33,21 +36,19 @@
...
     private final static String CONTACT_NOT_FOUND = "Contact not found in database";
+    private static final Logger LOGGER = LoggerFactory.getLogger("GsonDatabaseAdapter");
     private final String contactsPath;
     private final String profilesPath;
 
     /**
-     * Prevents construction from outside the class.
-     *
-     * @param serviceLocator Can be used to get access to other modules
+     * {@inheritDoc}
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: Can you revert this change ?

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java [hunk 4]
```diff
@@ -273,6 +274,7 @@ private void initializeDatabase() throws IOException {
     private void initializeContacts() throws IOException {
         final File file = new File(this.contactsPath);
         if (!file.exists()) {
+            LOGGER.warn("Database didn't exist -> created a new one");
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: Either move this log to the end of the if-statement or change `created a new one` to `creating a new one`. Also change `Database didn't exist` to `Contacts part of the database didn't exist` or move the log to the `initializeDatabase` method.

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java [hunk 5]
```diff
@@ -287,6 +289,7 @@ private void initializeContacts() throws IOException {
     private void initializeUserdata() throws IOException {
         final File file = new File(profilesPath);
         if (!file.exists()) {
+            LOGGER.warn("Database didn't exist -> created a new one");
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: Either move this log to the end of the if-statement or change `created a new one` to `creating a new one`. Also change `Database didn't exist` to `Userdata part of the database didn't exist` or move the log to the `initializeDatabase` method.

File: app/src/main/java/com/nervousfish/nervousfish/service_locator/EntryActivity.java [hunk 6]
```diff
@@ -21,6 +26,7 @@
     protected void onCreate(final Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
 
+        LOGGER.info("EntryActivity entered");
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/service_locator/EntryActivity.java]: Maybe change this to `EntryActivity created` and move it to the end of the method for clarity

File: app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java [hunk 7]
```diff
@@ -11,25 +11,29 @@
...
+    private static final Logger LOGGER = LoggerFactory.getLogger("LoginActivity");
     private EditText mPassword;
     private View mError;
     private IServiceLocator serviceLocator;
 
     @Override
     protected void onCreate(final Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+        setContentView(R.layout.login);
+        LOGGER.info("LoginActivity entered");
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java]: Maybe change this to `LoginActivity created` and move it to the end of the method for clarity

File: app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java [hunk 8]
```diff
@@ -38,6 +42,7 @@ protected void onCreate(final Bundle savedInstanceState) {
 
             @Override
             public void onClick(final View v) {
+                LOGGER.info("submitButton clicked");
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/LoginActivity.java]: While you're at it, maybe change this to `Submit button clicked`

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 9]
```diff
@@ -41,6 +45,7 @@
     protected void onCreate(final Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
+        LOGGER.info("MainActivity entered");
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Maybe change this to `MainActivity created` and move it to the end of the method for clarity

File: app/src/main/res/layout/content_contact.xml [hunk 10]
```diff
@@ -16,60 +36,37 @@
...
-                <TextView
-                    android:layout_width="wrap_content"
-                    android:layout_height="wrap_content"
-                    android:layout_marginTop="@dimen/contact_name_margin"
-                    android:text="@string/contact_title"
-                    android:textColor="@android:color/white"
-                    android:textSize="@dimen/contact_name_size"
-                    android:textStyle="bold" />
-
-            </LinearLayout>
```
Comment (ericcornelissen) [inline app/src/main/res/layout/content_contact.xml]: Are you aware that you ruined the layout due to these changes ?

Comment (jverbraeken) [inline app/src/main/res/layout/content_contact.xml]: Oops, fixed now
